### ISSUE TITLE
join-along throws exception

### ISSUE THREAD

**Issue by dmarjenburgh** [mikera/vectorz-clj#66]
The following code works for all built-in implementations of core.matrix:
```
(require '[clojure.core.matrix :as m])
(let [a (m/matrix [[1 3 5] [2 4 6]])
      b (m/matrix [7 8])]
  (m/join-along 1 a b))
=> [[1 3 5 7]
    [2 4 6 8]]
```
However, with the `:vectorz` implementation, I get the following exception:

```
IllegalArgumentException Incompatible shapes:
Matrix with shape [2,3] vs. Vector with shape [2] 
mikera.arrayz.impl.JoinedArray.join (JoinedArray.java:33)
```

**Comment by dmarjenburgh** [comment]
I just found out that the following *does* work:
```
(let [a (m/matrix [[1 3 5] [2 4 6]])
      b (m/matrix [[7] [8]])]
  (m/join-along 1 a b))
```
What is the expected way to deal with n-dimensional vectors as an nx1 matrix?

**Comment by mikera** [comment]
Hi,

`join-along` assumes that dimensions are compatible, i.e. it won't make assumptions for you like treating a N-vector as a Nx1 matrix. I'm actually a bit surprised it works with Clojure vectors... that should be undefined behaviour.

If you want to do this in general, you will need to specify how you want the vector to be reshaped, e.g.

```clojure
(let [a (matrix :vectorz [[1 3 5] [2 4 6]])
      b (matrix :vectorz [7 8])]
  (join-along 1 a (reshape b [2 1])))
=> #vectorz/array [[1.0,3.0,5.0,7.0],[2.0,4.0,6.0,8.0]]
```

**Comment by dmarjenburgh** [comment]
Would it be an idea to auto-reshape?
I expected to be able to join an `mxn` matrix M and a vector `v` in two ways:

- `(join-along 0 M v)` => (add v as row), shape goes from `[m n]` to `[m+1 n]`
- `(join-along 1 M v)` => (add v as column), shape goes from `[m n]` to `[m n+1]`

Generalizing to higher dimensions, there is a unique way to join two matrices `M` and `N` where `N` has one dimensionality less than `M`. Reshape `N` by inserting a 1 into the shape at the dimension you join along (E.g. the vector shape `[n]` would go to `[1 n]` as row-vector or `[n 1]` as column-vector).

Maybe this is the idea of `conjoin-along`, which tries to broadcast `N` before calling `join-along`, but it doesn't work for the example I posted.

**Comment by mikera** [comment]
That is indeed the general idea of `conjoin` and `conjoin-along`. It's effectively the array equivaleny of Clojure's `conj`, whereas `join` is the array equivalent of `concat`.

Might be a bug if it doesn't work? I shall have a quick dig.

**Comment by dmarjenburgh** [comment]
In that case, it's an issue in core.matrix. I can close this one and open an issue there.

**Comment by mikera** [comment]
Feel free to open a new issue.

Need to think about the edge cases though, e.g. should `conjoin-along` attempt to broadcast up to the desired shape? What happens when you try to conjoin a [5] vector to a [5 5 5] array for example? Should it broadcast to [5 5] and then insert another dimension to get [1 5 5], [5 1 5] or [5 5 1]?

**Comment by dmarjenburgh** [comment]
Yes, exactly. After conjoining a shape [5] vector to a shape [5 5 5] array, the resulting shape will be [6 5 5], [5 6 5] or [5 5 6] (according to the dimension you conjoin along).

In fact, you can get the added array with: `(slice result dim 5)`, since it's the last element of the slice across that dimension.

I got it working as intended by broadcasting first, then reshaping.

But then, the `:persistent-vector` and `:nd-array` implementations already work exactly like this. Only vectorz throws an exception.

**Comment by mikera** [comment]
Hmmmm broadcasting in core.matrix currently doesn't guarantee support for adding dimensions in non-leading positions. This means that [5] can broadcast to [1 1 5] but can't broadcast to [5 1 1] etc. which is probably why this is failing with vectorz-clj (which tends to be fairly strict on undefined behaviour...) Maybe we should revisit that assumption, it would certainly make this issue easier.

**Comment by dmarjenburgh** [comment]
I few solutions I see here:

- Loosen the constraint on join-along so the args don't need to have the same number of dimensions (it works this way for e.g. persistent-vector atm)
- Change conjoin-along to add the missing dimension of 1. Doesn't work with broadcast, so requires the added element to be a slice of the correct shape. (Breaks api)
- Change conjoin-along to broadcast to the shape of the slice is done now, but then reshape to add the extra dimension. This will work, but is a bit slower and requires both broadcast and reshape protocols to be implemented. This can be made more efficient by implementing a conjoin protocol (which is on the TODO list already)

**Comment by mikera** [comment]
I'm also interested in the idea of improving the broadcast operation to add missing dimensions. The problem with this though is that it makes broadcasting more expensive and more complex for implementators.

Maybe your suggestions make sense as an interim measure. 2 can be done more easily after broadcast changes?

**Comment by dmarjenburgh** [comment]
I don't know if that makes broadcasting a bit too complicated. I'm having a hard time understanding how that would work in the general case. Could you broadcast to an arbitrary shape?

I actually don't think broadcasting is broken or should be extended for this issue. Just conjoin-along should be fixed so you can add a slice along any dimension...

**Comment by mikera** [comment]
You can't broadcast to an arbitrary shape. 

**At present**: You can add arbitrary leading dimensions, so [5 5] can broadcast to [1 5 5] or [7 1 5 5] but not [5 1 5 or 5 1 5 7]

**Proposed**: You can add arbitrary leading dimensions and insert [1] dimensions anywhere, so [5 5] can broadcast to [1 5 5], [7 1 5 5] and [5 1 5] but not [5 1 5 7]

I believe the proposed solution gives a unique solution. If we were to relax the constraints further then I believe the result is ambiguous. e.g. when broadcasting a vector [3 5] to a shape of [2 2] it is ambiguous whether the result should be [[3 5] [3 5]] or [[3 3] [5 5]].

**Comment by dmarjenburgh** [comment]
Alright. Although I feel this route adds more complexity, I'd like to try and improve core.matrix. What are the steps needed for the extended broadcast operation to work for all implementations?

**Comment by dmarjenburgh** [comment]
I opened a pull request with a basic implementation.
https://github.com/mikera/core.matrix/pull/335
